druid-shell/src/platform/x11/xkb.rs

```diff
@@ -0,0 +1,456 @@
+// Copyright 2021 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A minimal wrapper around Xkb for our use.
+
+#![allow(non_upper_case_globals)]
+
+use super::xkbcommon_sys::*;
+use crate::{
+    platform::shared::{code_to_location, hardware_keycode_to_code},
+    KeyEvent, KeyState, Modifiers,
+};
+use keyboard_types::{Code, Key};
+use std::os::raw::{c_char, c_int};
+use std::{convert::TryFrom, ptr};
+use x11rb::xcb_ffi::XCBConnection;
+
+pub struct DeviceId(c_int);
+
+/// A global xkb context object.
+///
+/// Reference counted under the hood.
+// Assume this isn't threadsafe unless proved otherwise. (e.g. don't implement Send/Sync)
+// TODO do we need UnsafeCell?
```

Comment:
What would we need to check to find out? Is it a question of whether the FFI calls modify anything through the pointer?

Reply:
uh took code from wayland backend and missed this part :grimacing: . I will look into it

Reply:
I am not sure, why we need it; we only interact with `*mut xkb_context`

cc @derekdreery

Reply:
I've honestly forgot about how I did this with wayland, sorry!

Reply:
I don't think we need UnsafeCell, UnsafeCell is used to get `*mut T` from `&T`(well `&UnsafeCell<T>`)  but already have `*mut T`

Reply:
thought about this again, I have no doubts. 
I will merge in 2 days if no one objects.